### PR TITLE
Add Resampling 

### PR DESCRIPTION
**Pull request by tensionhead**
### Things added:
- backend resampling interfacing SciPy's resampling_poly
- backend tests
- frontend integration in /preproc/resampledata.py
### Still todo
- frontend tests

Author Guidelines
-----------------
- [x] Is the change set **< 400 lines**?
- [x] Was the code checked for memory leaks/performance bottlenecks?
- [x] Is the code running locally **and** on the ESI cluster?
- [x] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [ ] Are testing routines present?
- [ ] Do **parallel** loops have a set length and correct termination conditions?
- [ ] Do objects in the global package namespace perform proper parsing of their input? 
- [ ] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [ ] Is the code PEP8 compliant?
  - [ ] Does the code adhere to agreed-upon naming conventions?
  - [ ] Are keywords clearly named and easy to understand?
  - [ ] No commented-out code?
- [ ] Are all docstrings complete and accurate?

### CODE REVIEW

File: syncopy/preproc/resampling.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    """
+
+    # we need integers for slicing
+    skipped = int(samplerate // new_samplerate)
+
+    return dat[::skipped]
+
+
+def _get_updn(orig_fs, new_fs):
```
Comment (pantaray) [inline syncopy/preproc/resampling.py:121]: Using the `fractions` module for computing the sampling factors is really clever!

**Review by pantaray — COMMENTED** [review]
The backend interface looks very clean and robust. I anticipate some problems with the frontend, though (see comments). However, without tests, it's hard to say. One side-note: the current version of `local_spy.py` (here as well as in `master`) crashes with a `TypeError: only integer scalar arrays can be converted to a scalar index` raised in `synth_data.AR2_network`. Not sure if that's just a hiccup in the script or an actual bug.

File: syncopy/preproc/resampling.py (hunk 1 above)
Comment (tensionhead) [reply syncopy/preproc/resampling.py:121]: I think I got it from SO 😃 But yeah, `np.gcd` did not do the trick directly..

File: syncopy/preproc/resampledata.py [hunk 2]
```diff
@@ -11,74 +11,217 @@
...
+        scalar_parser(order, varname="order", lims=[0, np.inf], ntype="int_like")
+        if order < 100:
+            msg = ("You have chosen an anti-alias filter of very low "
+                   f"`order={order}`, expect a slow roll-off!"
+                   )
+            SPYWarning(msg)
+
+    # set default
+    else:
+        order = int(lenTrials.min()) if lenTrials.min() < 1000 else 1000
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/resampledata.py:124]: No idea :D Jokes aside: 1000ms recording with 1kHz sampling seems to me to be a standard trial size.. so no need to throw away filter sharpness here imho

[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/resampledata.py:124]: To elaborate a bit more: making the (default) filter larger than the signal has no effect, hence putting 5000 could mean for our average trial wasted computations 
I think something like an extreme lower-bound is maybe 80, but why not invest in a reasonable sized default filter which hopefully matches the trial sizes quite well?

**Comment by tensionhead** [comment]
> The backend interface looks very clean and robust. I anticipate some problems with the frontend, though (see comments). However, without tests, it's hard to say. One side-note: the current version of `local_spy.py` (here as well as in `master`) crashes with a `TypeError: only integer scalar arrays can be converted to a scalar index` raised in `synth_data.AR2_network`. Not sure if that's just a hiccup in the script or an actual bug.

I fixed it and wrote a little comment in the commit msg!

File: syncopy/preproc/compRoutines.py [hunk 3]
```diff
@@ -555,10 +556,7 @@ def process_metadata(self, data, out):
 
         # we need to re-calculate the downsampling factor
         factor = int(data.samplerate // self.cfg['new_samplerate'])
-        
-        # now set new samplerate
-        data.samplerate = self.cfg['new_samplerate']
-        
+
```
[the remark this answers is not in the document]
Comment (pantaray) [reply syncopy/preproc/compRoutines.py:559]: Okay - I trust you on the integer division, but the trial-codes in columns 4 and onwards will get mauled by `/factor`. Let's put a pin in this and keep it on our watch list.

File: syncopy/preproc/resampledata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pantaray) [reply syncopy/preproc/resampledata.py:124]: Got it, thanks for the clarification! I agree, 1000 seems like a good default for the kinds of data we're dealing with here.

File: syncopy/preproc/compRoutines.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/preproc/compRoutines.py:559]: I 'pinned' it in https://github.com/esi-neuroscience/syncopy/issues/290

**Pull request merged**